### PR TITLE
tic-tac-toe testing  #1

### PR DESCRIPTION
**Pull request by Elukoye**
This is the testing tic-tac-toe with the RSpec project. Kindly review.
Thanks

### CODE REVIEW

File: spec/board_spec.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+     expect(board).to eql([1,2,3,4,5,6,7,8,9])
+    end  
+  end 
+
+  
+  describe '#winner' do
+
+    let(:winner){@winner = nil}
+    it 'sets winner value to nil' do 
+      expect(winner).to eql(nil)
```
Comment (bolah2009) [inline spec/board_spec.rb]: :bug: This test is not connected with the Board Class, that is, commenting the code for Board Class still passes this test.

Let's use `let` to call an instance of the board class instead and test the methods on it.

#### This is true for all 15 `examples` in this test. Please do this for all other occurrences.

**Review by bolah2009 — CHANGES_REQUESTED** [review]
# Hi @Elukoye 

Nice job with the test setup :ok_hand: :smile: 

However, all test `examples` are redundant. :disappointed: Currently, the test is testing nothing about the Class.
- Let's instantiate the classes. e.g. `let(:board){Board.new}`, by doing this we are testing for the `#initialize` method an using the `Class` in the test.
- Let's call the public class methods e.g. `let(:win){board.win?}`, by doing this we are testing the methods of the initialized `Class`.
- Please add linter. Details can be found [here](https://github.com/microverseinc/linters-config/blob/master/ruby).
- Add a more discriptive readme. Details like how to run the test and [formatting the markdown file approprately](https://guides.github.com/features/mastering-markdown/) will make readme look professional.
- This is a collaborative project, is there a reason why you the only one working on this project? :thinking: 

Kindly, fix this and submit another code review request. :ok_hand: 

#### Feel free to reach me on slack if you having issues implementing these features. :+1: 

Good luck with your next submission. :star2: 

Cheers
[Bola](https://bolabuari.com/)

**Comment by Elukoye** [comment]
Made corrections as specified, kindly review.
Code syntax was checked locally.

File: spec/board_spec.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
+require './lib/board.rb'
+
+RSpec.describe Board do
+
+  let(:board){@board = Board.new}
```
Comment (bolah2009) [inline spec/board_spec.rb]: #### `REQUIRED`

Good job using `let`. :ok: Let's remove the `@board`, it's redundant in this case. See suggestion.

```suggestion
  let(:board){Board.new}
```

File: spec/board_spec.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+  describe '#display' do
+    output = 1..9
+    it 'displays the board ' do
+     expect(board.display).to eql(output)
+    end   
+  end 
+
+  describe '#position_taken?' do 
+    let(:board){@board = Board.new}
```
Comment (bolah2009) [inline spec/board_spec.rb]: There's no need to define `let(:board){@board = Board.new}` again in this case, as it is identical with the one defined above. The `let` is used to memorized helper methods and can be re-used.  :blush: See suggestion.

```suggestion
```

#### Learn more about [`let` here](https://relishapp.com/rspec/rspec-core/v/2-5/docs/helper-methods/let-and-let).

File: spec/board_spec.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    it 'checks if there is space on the board' do
+      expect(board.position_taken?(1)).to eql(false)
+    end
+  end
+
+  describe '#win?' do
+    let(:board){Board.new}
```
Comment (bolah2009) [inline spec/board_spec.rb]: See comment on re-using `let`.

```suggestion
```

File: spec/board_spec.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+  describe '#win?' do
+    let(:board){Board.new} 
+    it 'checks for win on board' do
+      expect(board.win?).to be false
+    end
+  end
+  
+  describe '#full?' do
+     let(:board){Board.new}
```
Comment (bolah2009) [inline spec/board_spec.rb]: See comment on re-using `let`.

```suggestion
```

File: spec/board_spec.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+     it 'checks if board is full' do
+      expect(board.full?).to eql(false)
+    end
+  end 
+
+  describe '#replace_var' do
+    let(:board){Board.new}
```
Comment (bolah2009) [inline spec/board_spec.rb]: See comment on re-using `let`.

```suggestion
```

File: spec/board_spec.rb [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    output = 1..9
+    it 'displays the board ' do
+     expect(board.display).to eql(output)
+    end   
+  end 
+
+  describe '#position_taken?' do 
+    let(:board){@board = Board.new}
+    it 'checks if there is space on the board' do
+      expect(board.position_taken?(1)).to eql(false)
```
Comment (bolah2009) [inline spec/board_spec.rb:16]: Let's test for both negative and positive cases.

```suggestion
      expect(board.position_taken?(1)).to eql(false)
      board.replace_var(0,'x')
      expect(board.position_taken?(1)).to eql(true)
```


 > Use your common sense: test positive and negative scenarios.

File: spec/board_spec.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+     let(:board){Board.new}
+     it 'checks if board is full' do
+      expect(board.full?).to eql(false)
+    end
+  end 
+
+  describe '#replace_var' do
+    let(:board){Board.new}
+    it 'replaces position with token on the board' do
+      expect(board.replace_var(1,'x')).to eql('x')
```
Comment (bolah2009) [inline spec/board_spec.rb]: Is this testing if the value have been really replaced? :thinking: 

Let's check the value of the board (`@board`) and the remaining value (`@remval`) too.

```suggestion
      expect(board.replace_var(1,'x')).to eql('x')
      expect(board.board).to eq [1, 'x', 3, 4, 5, 6, 7, 8, 9]
      expect(board.rem_nums).to eq [1, 3, 4, 5, 6, 7, 8, 9]
```

File: spec/board_spec.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    let(:board){@board = Board.new}
+    it 'checks if there is space on the board' do
+      expect(board.position_taken?(1)).to eql(false)
+    end
+  end
+
+  describe '#win?' do
+    let(:board){Board.new} 
+    it 'checks for win on board' do
+      expect(board.win?).to be false
```
Comment (bolah2009) [inline spec/board_spec.rb:27]: Let's add case when there's a winner. 
  > Use your common sense: test positive and negative scenarios.

The third line in the suggestion interately adds `x` to the board three times making `x` show appear in the first vertical line of the board, this make the `win?` the method detects the winner.

You can do for other positions.

```suggestion
      expect(board.win?).to be false
      (0..2).each { |i| board.replace_var(i, 'x') }
      expect(board.win?).to be true
```

File: spec/board_spec.rb [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+      expect(board.full?).to eql(false)
+    end
+  end 
+
+  describe '#replace_var' do
+    let(:board){Board.new}
+    it 'replaces position with token on the board' do
+      expect(board.replace_var(1,'x')).to eql('x')
+    end
+  end
```
Comment (bolah2009) [inline spec/board_spec.rb:83]: There's one more test missing. `#winner` method. Please let's add that too. An example can be like this:

```ruby
  describe '#winner' do
    it '...' do
      (0..2).each { |i| board.replace_var(i, 'x') }
      board.win?
      expect(board.winner).to be 'x'
    end
  end
```

**Review by bolah2009 — CHANGES_REQUESTED** [review]
# Nice Improvement :ok_hand: @Elukoye 

Some methods are not been tested... :blush: 

 - Please, let's add more test to the methods, also remember about negative and positive scenarios.
I added suggestions in the comments below.
- There are about `124 offenses detected` locally with `rubocop`. Please fix this. Also, check the [troubleshooting](https://github.com/microverseinc/linters-config/tree/master/ruby#troubleshooting) section for more details on how to fix the stickler checker not working. It should be displayed as shown below:
  succesful check PR

Kindly, fix this and submit another code review request. :ok_hand: 

#### Feel free to reach me on slack if you having issues implementing these features. :+1: 

Good luck with your next submission. :star2: 

Cheers
[Bola](https://bolabuari.com/)

File: spec/board_spec.rb (hunk 10 above)
Comment (Elukoye) [reply spec/board_spec.rb:83]: done

File: spec/board_spec.rb (hunk 9 above)
Comment (Elukoye) [reply spec/board_spec.rb:27]: done

File: spec/board_spec.rb (hunk 8 above)
Comment (Elukoye) [reply spec/board_spec.rb]: done

File: spec/board_spec.rb (hunk 7 above)
Comment (Elukoye) [reply spec/board_spec.rb:16]: done

**Comment by Elukoye** [comment]
@bolah2009 , kindly review. The specs have been tested locally and there is no error. Let me know if there is any feedback.

**Comment by Elukoye** [comment]
@bolah2009, I corrected some more specs, kindly review. All are tested locally and the tests have passed

**Review by Oghenebrume50 — CHANGES_REQUESTED** [review]
Hi @Elukoye  Good job working on the corrections so far :clap: everything looks good now, but stickler is not yet working on this repo yet. Have you activated it manually? Please try to do that submit for re-review. If you need help chat me up on slack

**Comment by Elukoye** [comment]
Enabled stickler in the repo. Kindly review

**Review by Oghenebrume50 — CHANGES_REQUESTED** [review]
Hi I see your progress, but stickler isn't enabled yet, have you switched it on for this repo from the stickler website? 

Did you visit this site? https://stickler-ci.com/teams/ If you have done all this just submit for a re-review and we would know it is an issue, else do it and fix the errors that show up if any.

**Comment by Elukoye** [comment]
I have enabled stickler in the repo ,I have included a stickler file in the
project ....so maybe you tell me what else to do ....or how to go about it .

**Comment by Elukoye** [comment]
Hi, I have checked the stickler website and it says it is enabled.

**Review by Oghenebrume50 — APPROVED** [review]
Hi, @Elukoye I see your progress. I do not think this is your fault,  these things happen sometimes, you may now submit this project, it is approved

**Comment by Elukoye** [comment]
Awesome thanks a mill.
